File: kirun-js/src/engine/function/system/math/RandomAny.ts
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../util/MapUtil';
import { AbstractFunction } from '../../AbstractFunction';

export class RandomAny extends AbstractFunction {
    public static readonly MIN_VALUE = 'minValue';

    public static readonly MAX_VALUE = 'maxValue';

    public static readonly VALUE = 'value';

    private readonly signature: FunctionSignature;

    private readonly randomFunction: (min: number, max: number) => number;

    public constructor(
        name: string,
        minParameter: Parameter,
        maxParameter: Parameter,
        outputSchema: Schema,
        randomFunction: (min: number, max: number) => number,
    ) {
        super();
        this.signature = new FunctionSignature(name)
            .setParameters(
                MapUtil.of(RandomAny.MIN_VALUE, minParameter, RandomAny.MAX_VALUE, maxParameter),
            )
            .setNamespace(Namespaces.MATH)
            .setEvents(
                new Map<string, Event>([
                    Event.outputEventMapEntry(MapUtil.of(RandomAny.VALUE, outputSchema)),
                ]),
            );
        this.randomFunction = randomFunction;
    }

    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let min: number = context.getArguments()?.get(RandomAny.MIN_VALUE);

        let max: number = context.getArguments()?.get(RandomAny.MAX_VALUE);

        let num: number = this.randomFunction(min, max);

        return new FunctionOutput([EventResult.outputOf(new Map([[RandomAny.VALUE, num]]))]);
    }
}

File: kirun-py/src/kirun_py/function/system/math/random_any.py
from __future__ import annotations

from typing import Callable, TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

MIN_VALUE = 'minValue'
MAX_VALUE = 'maxValue'
VALUE = 'value'


class RandomAny(AbstractFunction):

    def __init__(
        self,
        name: str,
        min_parameter: Parameter,
        max_parameter: Parameter,
        output_schema: Schema,
        random_function: Callable[[float, float], float],
    ) -> None:
        super().__init__()
        self._random_function = random_function
        self._signature = (
            FunctionSignature(name)
            .set_namespace(Namespaces.MATH)
            .set_parameters({
                MIN_VALUE: min_parameter,
                MAX_VALUE: max_parameter,
            })
            .set_events(
                dict([Event.output_event_map_entry({VALUE: output_schema})])
            )
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        args = context.get_arguments()
        min_val: float = args.get(MIN_VALUE)
        max_val: float = args.get(MAX_VALUE)
        num = self._random_function(min_val, max_val)
        return FunctionOutput([
            EventResult.output_of({VALUE: num})
        ])
